File: app/src/main/java/com/dmi/perfectreader/control/BookControl.java
package com.dmi.perfectreader.control;

import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;

import com.dmi.perfectreader.book.Book;
import com.dmi.perfectreader.main.GoBackIntent;
import com.dmi.perfectreader.main.ToggleMenuIntent;
import com.dmi.perfectreader.setting.Settings;
import com.dmi.perfectreader.util.android.EventBus;

import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EBean;

import static com.dmi.perfectreader.util.android.Units.dipToPx;
import static java.lang.Math.abs;
import static java.lang.Math.max;
import static java.lang.Math.min;
import static java.lang.Math.sqrt;

@EBean(scope = EBean.Scope.Singleton)
public class BookControl implements KeyEvent.Callback, View.OnTouchListener {
    private static final float TOUCH_SENSITIVITY = dipToPx(8);
    private static final float LEFT_SIDE_WIDTH_FOR_SLIDE = dipToPx(40);
    private static final float SLIDE_SENSITIVITY = dipToPx(20);

    private static final int FONT_SIZE_MAX = 800;
    private static final int FONT_SIZE_MIN = 20;
    private static final int FONT_SIZE_DELTA = 10;

    @Bean
    protected EventBus eventBus;
    @Bean
    protected Settings settings;
    private Book book;

    private int width;
    private int height;
    private float touchDownX;
    private float touchDownY;
    private float oldApplySlideActionTouchY;
    private boolean nowIsSlideByLeftSide = false;

    public void setBook(Book book) {
        this.book = book;
    }

    @Override
    public boolean onTouch(View view, MotionEvent event) {
        width = view.getWidth();
        height = view.getHeight();
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                onTouchDown(event);
                break;
            case MotionEvent.ACTION_MOVE:
                onTouchMove(event);
                break;
            case MotionEvent.ACTION_UP:
                onTouchUp(event);
                break;
        }
        return true;
    }

    private void onTouchDown(MotionEvent motionEvent) {
        touchDownX = motionEvent.getX();
        touchDownY = motionEvent.getY();
        oldApplySlideActionTouchY = touchDownY;
        nowIsSlideByLeftSide = false;
    }

    private void onTouchMove(MotionEvent motionEvent) {
        boolean onLeftSide = touchDownX <= LEFT_SIDE_WIDTH_FOR_SLIDE;
        boolean isTouchedFar = abs(motionEvent.getY() - touchDownY) >= TOUCH_SENSITIVITY;
        if (onLeftSide && isTouchedFar) {
            nowIsSlideByLeftSide = true;
        }

        if (nowIsSlideByLeftSide) {
            if (abs(oldApplySlideActionTouchY - motionEvent.getY()) >= SLIDE_SENSITIVITY) {
                int count = (int) ((motionEvent.getY() - oldApplySlideActionTouchY) / SLIDE_SENSITIVITY);
                int fontSize = settings.format.fontSize.get();
                fontSize = max(FONT_SIZE_MIN, min(FONT_SIZE_MAX, fontSize + count * FONT_SIZE_DELTA));
                settings.format.fontSize.set(fontSize);
                book.configure().setFontSize(fontSize).commit();
                oldApplySlideActionTouchY = motionEvent.getY();
            }
        }
    }

    private void onTouchUp(MotionEvent motionEvent) {
        if (!nowIsSlideByLeftSide) {
            float touchOffsetX = motionEvent.getX() - touchDownX;
            float touchOffsetY = motionEvent.getY() - touchDownY;
            float touchOffset = (float) sqrt(touchOffsetX * touchOffsetX + touchOffsetY * touchOffsetY);
            boolean isTap = touchOffset <= TOUCH_SENSITIVITY;

            if (isTap) {
                float xPart = motionEvent.getX() / width;
                float yPart = motionEvent.getY() / height;
                TapZoneConfiguration configuration = settings.control.tapZones.shortTaps.configuration.get();
                TapZone tapZone = configuration.getAt(xPart, yPart);
                Action action = settings.control.tapZones.shortTaps.action(tapZone).get();
                performAction(action);
            } else {
                boolean swipeLeft = touchOffsetX <= -TOUCH_SENSITIVITY;
                boolean swipeRight = touchOffsetX >= TOUCH_SENSITIVITY;
                if (swipeLeft) {
                    book.goNextPage();
                } else if (swipeRight) {
                    book.goPreviewPage();
                }
            }
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        HardKey hardKey = HardKey.fromKeyCode(keyCode);
        if (hardKey != HardKey.UNKNOWN) {
            Action action = settings.control.hardKeys.shortPress.action(hardKey).get();
            performAction(action);
        }
        return true;
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        return true;
    }

    @Override
    public boolean onKeyLongPress(int keyCode, KeyEvent event) {
        return false;
    }

    @Override
    public boolean onKeyMultiple(int keyCode, int count, KeyEvent event) {
        return false;
    }

    private void performAction(Action action) {
        switch (action) {
            case NONE:
                break;
            case TOGGLE_MENU:
                eventBus.post(new ToggleMenuIntent());
                break;
            case GO_BACK:
                eventBus.post(new GoBackIntent());
                break;
            case GO_NEXT_PAGE:
                book.goNextPage();
                break;
            case GO_PREVIEW_PAGE:
                book.goPreviewPage();
                break;
            case SELECT_TEXT:
                book.selectText();
                break;
            default:
                throw new UnsupportedOperationException();
        }
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/fragment/bookcontrol/BookControl.kt
package com.dmi.perfectreader.fragment.bookcontrol

import com.dmi.perfectreader.fragment.book.Book
import com.dmi.perfectreader.fragment.book.location.LocationRange
import com.dmi.perfectreader.fragment.book.selection.selectionCharAt
import com.dmi.perfectreader.fragment.bookcontrol.entity.Action
import com.dmi.perfectreader.fragment.bookcontrol.entity.ControlMode
import com.dmi.perfectreader.fragment.bookcontrol.entity.HardKey
import com.dmi.perfectreader.fragment.bookcontrol.entity.TouchInfo
import com.dmi.perfectreader.fragment.reader.Reader
import com.dmi.util.android.base.BaseViewModel
import com.dmi.util.graphic.SizeF
import com.dmi.util.setting.Settings
import com.dmi.perfectreader.data.UserSettingKeys.Control as ControlKeys
import com.dmi.perfectreader.data.UserSettingKeys.Format as FormatKeys

class BookControl(
        private val userSettings: Settings,
        private val book: Book,
        private val reader: Reader,
        private val closeApp: () -> Unit,
        dip2px: (Float) -> Float
) : BaseViewModel() {
    private val TOUCH_SENSITIVITY = dip2px(8F)
    private val LEFT_SIDE_WIDTH_FOR_SLIDE = dip2px(40F)
    private val SLIDE_SENSITIVITY = dip2px(20F)

    private val FONT_SIZE_MAX = 4F
    private val FONT_SIZE_MIN = 0.2F
    private val FONT_SIZE_DELTA = 0.05F

    private lateinit var size: SizeF

    private var mode: ControlMode by saveState(ControlMode.NORMAL)
    private var touchDownX = 0F
    private var touchDownY = 0F
    private var oldApplySlideActionTouchY = 0F
    private var nowIsSlideByLeftSide = false

    fun resize(size: SizeF) = run { this.size = size }

    fun onTouchDown(touchInfo: TouchInfo) {
        touchDownX = touchInfo.x
        touchDownY = touchInfo.y
        oldApplySlideActionTouchY = touchDownY
        nowIsSlideByLeftSide = false
    }

    fun onTouchMove(touchInfo: TouchInfo) {
        val y = touchInfo.y
        val onLeftSide = touchDownX <= LEFT_SIDE_WIDTH_FOR_SLIDE
        val isTouchedFar = Math.abs(y - touchDownY) >= TOUCH_SENSITIVITY
        if (onLeftSide && isTouchedFar) {
            nowIsSlideByLeftSide = true
        }

        if (nowIsSlideByLeftSide) {
            if (Math.abs(oldApplySlideActionTouchY - y) >= SLIDE_SENSITIVITY) {
                val count = ((y - oldApplySlideActionTouchY) / SLIDE_SENSITIVITY).toInt()
                changeFontSize(count)
                oldApplySlideActionTouchY = y
            }
        }
    }

    private fun changeFontSize(count: Int) {
        val fontSize = userSettings[FormatKeys.fontSizeMultiplier]
        val newFontSize = Math.max(FONT_SIZE_MIN, Math.min(FONT_SIZE_MAX, fontSize + count * FONT_SIZE_DELTA))
        if (newFontSize != fontSize) {
            userSettings[FormatKeys.fontSizeMultiplier] = newFontSize
            book.reformat()
        }
    }

    fun onTouchUp(touchInfo: TouchInfo) {
        val x = touchInfo.x
        val y = touchInfo.y
        val radius = touchInfo.radius
        if (!nowIsSlideByLeftSide) {
            val touchOffsetX = x - touchDownX
            val touchOffsetY = y - touchDownY
            val touchOffset = Math.sqrt((touchOffsetX * touchOffsetX + touchOffsetY * touchOffsetY).toDouble()).toFloat()
            val isTap = touchOffset <= TOUCH_SENSITIVITY

            if (isTap) {
                if (mode == ControlMode.NORMAL) {
                    // todo убрать после тестирования выделения текста
//                    val xPart = x / size.width
//                    val yPart = y / size.height
//                    val configuration = userSettings[ControlKeys.TapZones.ShortTaps.configuration]
//                    val tapZone = configuration.getAt(xPart, yPart)
//                    val action = userSettings[ControlKeys.TapZones.ShortTaps.Actions[tapZone]]
//                    performCoordinatedAction(action, x, y, radius)
                    startSelection(x, y)
                } else {
                    cancelSelection()
                }
            } else {
                val swipeLeft = touchOffsetX <= -TOUCH_SENSITIVITY
                val swipeRight = touchOffsetX >= TOUCH_SENSITIVITY
                if (swipeLeft) {
                    book.goNextPage()
                } else if (swipeRight) {
                    book.goPreviousPage()
                }
            }
        }
    }

    fun onKeyDown(hardKey: HardKey) {
        val action = userSettings[ControlKeys.HardKeys.ShortPress.Actions[hardKey]]
        performAction(action)
    }

    private fun performCoordinatedAction(action: Action, x: Float, y: Float, radius: Float) = when (action) {
        Action.SELECT_TEXT -> startSelection(x, y)
        else -> performAction(action)
    }

    private fun performAction(action: Action) = when (action) {
        Action.NONE -> Unit
        Action.TOGGLE_MENU -> reader.toggleMenu()
        Action.EXIT_APP -> closeApp()
        Action.GO_NEXT_PAGE -> book.goNextPage()
        Action.GO_PREVIOUS_PAGE -> book.goPreviousPage()
        else -> throw IllegalStateException("Action didn't handle")
    }

    private fun startSelection(x: Float, y: Float) {
        val currentPage = book.pageAt(0)
        if (currentPage != null) {
            val charUnderTouch = selectionCharAt(currentPage, x, y)
            if (charUnderTouch != null) {
                val begin = charUnderTouch.obj.charLocation(charUnderTouch.charIndex)
                val end = charUnderTouch.obj.charLocation(charUnderTouch.charIndex + 1)
                book.selectionRange = LocationRange(begin, end)

                reader.toggleSelection()
                mode = ControlMode.SELECTION
            }
        }
    }

    private fun cancelSelection() {
        reader.toggleSelection()
        mode = ControlMode.NORMAL
    }
}